pkg/ipcache/types.go

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ipcache
+
+import (
+	"github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+// prefixInfo holds all of the information (labels, etc.) about a given prefix
+// independently based on the ResourceID of the origin of that information, and
+// provides convenient accessors to consistently merge the stored information
+// to generate ipcache output based on a range of inputs.
+type prefixInfo map[types.ResourceID]*resourceInfo
+
+// resourceInfo is all of the information that has been collected from a given
+// resource (types.ResourceID) about this IP. Each field must have a 'zero'
+// value that indicates that it should be ignored for purposes of merging
+// multiple resourceInfo across multiple ResourceIDs together.
+type resourceInfo struct {
+	labels labels.Labels
+	source source.Source
+}
+
+// IPMetadata is an empty interface intended to inform developers using the
+// IPCache interface about which types are valid to be injected, and how to
+// update this code, in particular the merge(),unmerge(),isValid() methods
+// below.
+//
+// In an ideal world, we would use Constraints here but as of Go 1.18, these
+// cannot be used in conjunction with methods, which is how the information
+// gets injected into the IPCache.
+type IPMetadata interface{}
+
+// merge overwrites the field in 'resourceInfo' corresponding to 'info'. This
+// associates the new information with the prefix and ResourceID that this
+// 'resourceInfo' resides under in the outer metadata map.
+func (m *resourceInfo) merge(info IPMetadata, src source.Source) {
+	switch info := info.(type) {
+	case labels.Labels:
+		l := labels.NewLabelsFromModel(nil)
+		l.MergeLabels(info)
+		m.labels = l
+	default:
+		log.Errorf("BUG: Invalid IPMetadata passed to ipinfo.merge(): %+v", info)
```

Comment:
What I would _like_ here is actually a compile-time failure. I'm not sure the best way to do this though.. I tried playing around with generics but that wouldn't work because these are methods defined against some receiver, not functions. Go generics only work for functions AFAIK.

I think you're probably right that it's better to avoid merging any data here in case it impacts something else without actually implementing this merge. :+1: I'll add a return statement here.